src/include/miopen/env.hpp

```diff
@@ -31,127 +31,166 @@
-// Supports hexadecimal with leading 0x or decimal
-inline uint64_t EnvvarValue(const char* name, uint64_t fallback = 0)
+template <>
+struct ParseEnvVal<uint64_t>
 {
-    // NOLINTNEXTLINE (concurrency-mt-unsafe)
-    const auto value_env_p = std::getenv(name);
-    if(value_env_p == nullptr)
+    static uint64_t go(const char* vp) { return std::strtoull(vp, nullptr, 0); }
+};
+
+template <>
+struct ParseEnvVal<std::string>
+{
+    static std::string go(const char* vp) { return std::string{vp}; }
+};
+
+template <typename T>
+struct EnvVar
+{
+private:
+    T value{};
+    bool is_default = true;
+
+public:
+    const T& GetValue() const { return value; }
+
+    bool IsDefault() const { return is_default; }
+
+    void UpdateValue(const T& val)
     {
-        return fallback;
+        is_default = false;
+        value      = val;
     }
-    else
+
+    explicit EnvVar(const char* const name, const T& def_val)
     {
-        return strtoull(value_env_p, nullptr, 0);
+        // NOLINTNEXTLINE (concurrency-mt-unsafe)
+        const char* vp = std::getenv(name);
+        if(vp != nullptr) // a value was provided
+        {
+            is_default = false;
+            value      = ParseEnvVal<T>::go(vp);
+        }
+        else // no value provided, use default value
+        {
+            value = def_val;
+        }
     }
+};
+
+} // end namespace internal
+
+// static inside function hides the variable and provides
+// thread-safety/locking
+#define MIOPEN_DECLARE_ENV_VAR(name, type, default_val)                    \
+    struct name                                                            \
+    {                                                                      \
+        using value_type = type;                                           \
+        static miopen::internal::EnvVar<type>& Ref()                       \
+        {                                                                  \
+            static miopen::internal::EnvVar<type> var{#name, default_val}; \
+            return var;                                                    \
+        }                                                                  \
+    };
+
+/// \todo the following functions should be renamed to either include the word Env
+/// or put inside a namespace 'env'. Right now we have a function named Value()
+/// that returns env var value as only 64-bit ints
+
+template <class EnvVar>
+inline const std::string& GetStringEnv(EnvVar)
+{
+    static_assert(std::is_same_v<typename EnvVar::value_type, std::string>);
+    return EnvVar::Ref().GetValue();
 }
 
-inline std::vector<std::string> GetEnv(const char* name)
+template <class EnvVar>
+inline bool IsEnabled(EnvVar)
 {
-    // NOLINTNEXTLINE (concurrency-mt-unsafe)
-    const auto p = std::getenv(name);
-    if(p == nullptr)
-        return {};
-    else
-        return {{p}};
+    static_assert(std::is_same_v<typename EnvVar::value_type, bool>);
+    return EnvVar::Ref().GetValue();
```

Comment:
```suggestion
    return !EnvVar::Ref().IsUnset() && EnvVar::Ref().GetValue();
```

Reply:
@atamazov : I am against introducing the concept of `Set/Unset`. The whole idea with `default_value` is to avoid this altogether. Any uses of `environment variable` should not need to check if the variable was provided a value because it will always have the default value. 

Now the `IsDefault/is_default` currently is a hack put in place to retain some tricky existing behaviors that are dependent on `env var` being initialized differently under different scenarios and `IsEnabled/IsDisabled` alone do not preserve the existing behavior. 

So to simplify our future life, we should deal with those tricky scenarios one by one (i.e. all the places that use `IsDefault()` now) and then we get rid of `IsDefault()` as well. 

CC: @JehandadKhan .

Reply:
@amberhassaan Thanks for your opinion. However my comment is addressed to the author or this PR, @cderb. Let's listen to what he has to say.

Reply:
@atamazov : I designed this logic in `env.hpp` and handed it over to @cderb, resulting from the need for updating the cached values of environment variables for our `gtests`. @JehandadKhan was also involved in the discussion, hence I am commenting and I firmly believe that besides the updatable functionality, this design is an improvement over the existing design due to various reasons mentioned in my comments.

Reply:
Good observation, this does restore the original functionality of IsDisabled / IsEnabled.

I'm attempting to weigh these 2 schools of thought:
@amberhassaan believes that an env flag should match to a boolean var. This var should have a default value, specified at instantiation. This would mean IsEnabled == !IsDisabled is always true (no purpose in having both). This would also mean that env vars function more akin to global variables.

@atamazov would like to keep the behavior in which IsEnabled reflects explicit enablement (env was set) by end user, as well as IsDisabled reflects explicit disablement by end user. In which IsEnabled == !IsDisabled is only true when the env has been set. In this sense the function used determines the default behavior.

I believe there is some merit in distinguishing that an environment wasn't set by the end user.
I'll draft a change with this in mind and see how it looks.

Reply:
@amberhassaan 
> @atamazov : I designed this logic in `env.hpp` and handed it over to @cderb, resulting from the need for updating the cached values of environment variables for our `gtests`. @JehandadKhan was also involved in the discussion...

Good! I can also recommend discussing changes with the authors of the original code, especially if the changes are debatable.

Reply:
> 
> I believe there is some merit in distinguishing that an environment wasn't set by the end user. I'll draft a change with this in mind and see how it looks.

@cderb : Can you please put a `todo` in all the places where you had to use `IsDefault()` previously, so that we can go back and simplify that tricky logic, especially if we decide to go with true boolean semantics. Given your choice of Set/UnSet, I fear we'll lose those spots and it'll be hard to find them again.

Reply:
I do believe that readability of code is equally if not more important than
* reducing the number of changes
* how we perceive good code should look like 
I find our existing method for testing env vars as well as knowning what the default value is. Therefore, the code as it currently exists in this PR would be my preference. Since it make it clear what the default value is and also in the code branching what are we checking. Furthermore, to my mind an `if` clause with a single predicate is much more preferable compraed to a complicated combination of various things that needs to be figured out to determine how the code would behave. 

Just my two cents for the PR author to consider ;)

Reply:
@JehandadKhan 
> Furthermore, to my mind an if clause with a single predicate is much more preferable compared to a complicated combination of various things that needs to be figured out to determine how the code would behave.

Just to clarify: do you prefer stuff like 
```C++
if(miopen::IsDefault(MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_ASM_FWD_GTC_XDLOPS{}) ||
           !miopen::IsEnabled(MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_ASM_FWD_GTC_XDLOPS{}))
```
in many places over
```C++
return !EnvVar::Ref().IsUnset() && !EnvVar::Ref().GetValue();
```
in one place?

Reply:
I am of the opinion that we should have neither. We should choose `default values` and design logic such that it does not matter whether the variable was `set`. I think it's quite possible to do so and will lead to simpler code.

Reply:
@atamazov  I agree with @amberhassaan, with the default value specified in the macro and the predicate just checking whether the value of the env var is default or not is much easier to follow.

Reply:
@JehandadKhan @amberhassaan Just to make sure that you understand that the original code implements and uses the tri-state object {UNSET, TRUE, FALSE}. This is not a boolean. The unset state is a natural property of environment variables, it is widely used in the library and has proven to be useful.

<details><summary><b><i>EXAMPLE</i></b></summary><hr>

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/c87e82af6b6c5e487f59c6fa4d85c353900ab4a1/src/solver/conv_hip_implicit_gemm_bwd_v4r1_xdlops.cpp#L813-L831
<hr></details>

In order to cut this feature out of the library, we need more serious reasons than what I see so far.

Reply:
What is the advantage of treating env vars as tri-state ? Given that similar functionality may be achieved by using a known initialization variable ? What would we loose ?

Reply:
@atamazov : simplicity is a good enough reason. @cderb showed me this code (linked below) and it took us a day of discussing and understanding the intended behavior. This is not how we should be using our env flags: 
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/c87e82af6b6c5e487f59c6fa4d85c353900ab4a1/src/solver/conv_ocl_dir2D_bwdWrW_1x1.cpp#L48-L55

Reply:
I will also add the MIOpen is the only library that I've come across (in my limited experience) that treats env flags like this.

Reply:
@amberhassaan 
> This is not how we should be using our env flags...

No problem. Just show your proposal (code snippet) for replacement. Or just make the fixes you prefer. I am not going to block your changes. 

> @cderb showed me this code (linked below) and it took us a day of discussing and understanding the intended behavior. 

💡 git blame -> #1329 -> "You can use the above ENV vars to override (i.e. enable) OCL and HIP convolutions."

Reply:
Looking at the PR gives context but I still find the logic hard to understand, especially now that I know that there are 3 different code paths possible because of the `tri-state` nature of env flags, and that `Enabled != !Disabled`. I think of these as switches `on/off` and find `on/off` simpler to understand than `tri-state`. 

I'll put out a PR when I get around to it.